src/wp-includes/blocks.php

```diff
@@ -760,6 +760,86 @@ function get_hooked_blocks( $name ) {
 	return $hooked_blocks;
 }
 
+/**
+ * Returns a function that...
+ *
+ * @since 6.4.0
+ * @access private
+ *
+ * @param WP_Block_Template|array $context A block template, template part, or pattern that the blocks belong to.
+ * @return callable A function that returns the serialized markup for the given block.
+ */
+function make_before_block_visitor( $context ) {
+	return function( &$block, $parent = null, $prev = null ) use ( $context ) {
+		_inject_theme_attribute_in_template_part_block( $block );
+
+		$markup = '';
+
+		if ( $parent && ! $prev ) {
+			// Candidate for first-child insertion.
+			$hooked_blocks_for_parent = get_hooked_blocks( $parent['blockName'] );
+			foreach ( $hooked_blocks_for_parent as $hooked_block_type => $relative_position ) {
+				if ( 'first_child' === $relative_position ) {
+					$hooked_block_markup = sprintf( '<!-- wp:%s /-->', strip_core_block_namespace( $hooked_block_type ) );
+					/** This filter is documented in wp-includes/blocks.php */
+					$markup .= apply_filters( 'insert_hooked_block', $hooked_block_markup, $hooked_block_type, $relative_position, $parent, $context );
+				}
+			}
+		}
+
+		$hooked_blocks = get_hooked_blocks( $block['blockName'] );
+		foreach ( $hooked_blocks as $hooked_block_type => $relative_position ) {
+			if ( 'before' === $relative_position ) {
+				$hooked_block_markup = sprintf( '<!-- wp:%s /-->', strip_core_block_namespace( $hooked_block_type ) );
+				/**
+				 * Filters the serialized markup of a hooked block.
+				 *
+				 * @since 6.4.0
+				 *
+				 * @param string                  $hooked_block_markup The serialized markup of the hooked block.
+				 * @param string                  $hooked_block_type   The type of the hooked block.
+				 * @param string                  $relative_position   The relative position of the hooked block.
+				 *                                                     Can be one of 'before', 'after', 'first_child', or 'last_child'.
+				 * @param array                   $block               The anchor block.
+				 * @param WP_Block_Template|array $context             The block template, template part, or pattern that the anchor block belongs to.
+				 */
+				$markup .= apply_filters( 'insert_hooked_block', $hooked_block_markup, $hooked_block_type, $relative_position, $block, $context );
```

Comment:
This is nice, so you essentially do:

```php
add_filter( 'insert_hooked_block', function( $markup, $block_type, $position ) {
    if ( 'tests/some-block' === $block_type && 'after' === $position ) {
        return '';
    }
    return $markup;
}, 10, 3 );
```

In case you want to prevent injecting a block matching your conditions. It can also cover the current block and the context: pattern or template. Powerful!

Reply:
Exactly, that's the idea!

One thing we might still want to add (in a follow-up) could be a similar filter to allow _adding_ a block conditionally 🤔 (So it doesn't need to be registered in a `block.json` first.)